File: python/test/02-builtins.py
#!/bin/bash

##############################################################################
# BOOTSTRAP
#
# Use python3 or python, whichever is available.
# (See https://unix.stackexchange.com/questions/20880)
#
if "true" : '''\'
then
    exec $(command -v python3 || command -v python) "$0" "$@"
    exit 127
fi
'''


##############################################################################
# PYTHON CODE BEGINS HERE

import sys
import liteexpr

BUILTINS = [
    "CEIL",
    "EVAL",
    "FLOOR",
    "FOR",
    "FOREACH",
    "FUNCTION",
    "IF",
    "LEN",
    "PRINT",
    "ROUND",
    "SQRT",
    "WHILE",
]


def main():
    nan = float("NaN")
    inf = float("Inf")
    ninf = float("-Inf")
    grades = liteexpr.SymbolTable({
        "grades" : {
            "Alice"   : "A",
            "Bob"     : "B",
            "Charlie" : "C"
        }
    })

    for i in [ 0, 7, 3.5, -7, -3.5, nan, inf, ninf, '""', '"x"' ]:
        for builtin in BUILTINS:
            symbols = liteexpr.SymbolTable({ "i":i })
            expr = f"{builtin}(i)"

            try:
                result = liteexpr.eval(expr, symbols)
            except liteexpr.Error as e:
                result = str(e).replace("\n", "")

            expr = f"{builtin}({i})"
            print(f"{expr:<15} => {result}")

    result = liteexpr.eval("""
        FOREACH(record, grades,
            name = record[0];
            grade = record[1];

            PRINT(name + "'s grade is " + grade);
        )
    """, grades)

    print(f"=> {result}")


##############################################################################
# ENTRY POINT

if __name__ == "__main__":
    main()


# vim:filetype=python:

File: cpp/test/02-builtins.cpp
#include <vector>
#include <string>
#include <iostream>
#include "liteexpr.h"

using namespace std;

vector<string> BUILTINS = {
    "CEIL",
    "EVAL",
    "FLOOR",
    "FOR",
    "FOREACH",
    "FUNCTION",
    "IF",
    "LEN",
    "PRINT",
    "ROUND",
    "SQRT",
    "WHILE",
};


const double NaN = 0.0/0.0;
const double Inf = 1.0/0.0;


int main(int argc, const char* argv[]) {
    string result;
    liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
        { "grades", liteexpr::make_value({
            { "Alice"   , liteexpr::make_value("A") },
            { "Bob"     , liteexpr::make_value("B") },
            { "Charlie" , liteexpr::make_value("C") },
        })},
    });
    liteexpr::ARRAY list = liteexpr::make_array({
        liteexpr::make_value(0),
        liteexpr::make_value(7),
        liteexpr::make_value(3.5),
        liteexpr::make_value(-7),
        liteexpr::make_value(-3.5),
        liteexpr::make_value(NaN),
        liteexpr::make_value(Inf),
        liteexpr::make_value(-Inf),
        liteexpr::make_value("\"\""),
        liteexpr::make_value("\"x\""),
    });

    for(liteexpr::VALUE i : list->native()) {
        for(string builtin : BUILTINS) {
            string expr = builtin + "(i)";
            string pexpr = builtin + "(" + i->svalue() + ")";
            liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
                { "i", i },
            });

            try {
                result = liteexpr::eval(expr, symbols)->svalue();
            }
            catch(liteexpr::Error e) {
                result = string(e);
                replace(result.begin(), result.end(), '\n', '\t');
            }

            cout << setw(15) << left << pexpr << " => " << result << endl;
        }
    }

    try {
        result = liteexpr::eval(R"(
            FOREACH(record, grades,
                name = record[0];
                grade = record[1];

                PRINT(name + "'s grade is " + grade);
            )
        )", symbols)->svalue();
    }
    catch(liteexpr::Error e) {
        result = string(e);
        replace(result.begin(), result.end(), '\n', '\t');
    }

    cout << "=> " << result << endl;

    return 0;
}
